patterns/multitenancy.md

```diff
@@ -0,0 +1,180 @@
+# Multi-tenancy API v1
+
+
+## Context
+
+Fragment uit de offerte vraag wat betreft het bouwen van een engine:
+“Een engine is multi-tenant by design. Dit betekent dat één instantie van de software oplossing meerdere tenants kan voorzien. Elke tenant is een afzonderlijk gescheiden logische omgeving, elk met zijn eigen set gebruikers, autorisatie regels en data.”
+
+
+## Conceptueel
+
+Multi-tenancy via twee hoedanigheden : providers en consumers.
+
+![multitenancy_conceptueel](/img/multitenancy_conceptueel_20180313.jpg)
+
+## Terminologie
+
+| Term | Definitie |
+| :--- | :--- |
+| Multi-tenant | Een component is multi-tenant indien één instantie van de software oplossing meerdere tenants kan voorzien. Elke tenant is een afzonderlijk gescheiden logische omgeving, elk met zijn eigen set gebruikers, autorisatie regels en data. Een service is multi-tenant indien deze de multi-tenancy API implementeert. |
+| Multi-instance| Een uitbreiding van een multi-tenant component (zie boven) waarbij men gebruik maakt van afzonderlijk gescheiden processen. |
+| Tenant-aware | Een client component is tenant-aware indien deze expliciet een ‘tenant id’ gebruikt bij het consumeren van een andere component. |
+| Tenant catalog | Een component draagt zelf de verantwoordelijkheid voor persistentie van de tenant id’s, dit geldt voor de provider of consumer kant, of beide. De persistentie gebeurd in de zogenaamde 'tenant catalog'. |
+| Tenant provisioning | De Multi-tenant API voorziet beheer methoden om nieuwe tenant id’s en mapping te provisioneren. | 
+| Tenant id | Dit is de unieke identificatie van een tenant binnen een bepaalde component.
+
+
+## Assumpties
+
+* Elke (nieuwe) (ACPaaS) component wordt verwacht rekening te houden met multi-tenancy, hetzij in de hoedanigheid van provider of als consumer. 
+Meer specifiek: elke toekomstige engine is multi-tenant (cfr. offerte vraag), maw elke nieuwe engine is een provider.
+
+* Er is geen centraal tenant beheersysteem.
+De provider componenten zelf zijn verantwoordelijk voor tenant beheer en persistentie. In de praktijk 1 of meerdere tenants. Elke tenant heeft  een uniek tenant id. 
```

Comment:
... zijn zelf verantwoordelijk ... In de praktijk heeft men altijd 1 of meerder tenants.